File: godot_yaml_rust/src/godot/variant.cpp
#include "godot/string.h" // NOLINT(modernize-deprecated-headers)
#include "core/Variant.hpp"
#include "core/String.hpp"

namespace gcs {
    namespace ffi {
        int64_t variant_as_i64(const ::godot::Variant &variant) {
            return variant.operator int64_t();
        }

        ::rust::string variant_as_string(const ::godot::Variant &variant) {
            auto as_string = variant.operator ::godot::String();
            return godot_string_to_rust_string(as_string);
        }

        bool variant_as_bool(const ::godot::Variant &variant) {
            return variant.operator bool();
        }

        double variant_as_f64(const ::godot::Variant &variant) {
            return variant.operator double();
        }

        const ::godot::Variant &empty_variant() {
            return *(new ::godot::Variant());
        }

        const ::godot::Variant &variant_from_i64(int64_t value) {
            return *(new ::godot::Variant(value));
        }

        const ::godot::Variant &variant_from_string(::rust::string value) {
            return *(new ::godot::Variant(string_from_rust_string(value)));
        }

        const ::godot::Variant &variant_from_bool(bool value) {
            return *(new ::godot::Variant(value));
        }

        const ::godot::Variant &variant_from_f64(double value) {
            return *(new ::godot::Variant(value));
        }
    }
}

File: godot_yaml_rust/src/godot/variant.rs
use crate::godot::variant::ffi::Variant;
use cxx::{type_id, ExternType};
use godot::variant::VariantType;

#[cxx::bridge(namespace = godot::ffi)]
pub mod ffi {
    unsafe extern "C++" {
        include!("godot_yaml_rust/include/godot/variant.h");
        include!("rust/cxx.h");
        pub type Variant;
        #[cxx_name = "VariantType"]
        pub type CXXVariantType = crate::godot::variant::CXXVariantType;

        pub fn get_type(self: &Variant) -> CXXVariantType;

        pub fn variant_as_i64(variant: &Variant) -> i64;
        pub fn variant_as_string(variant: &Variant) -> String;
        pub fn variant_as_bool(variant: &Variant) -> bool;
        pub fn variant_as_f64(variant: &Variant) -> f64;
    }
}

#[derive(Hash, Eq, PartialEq, Clone, Default)]
pub struct CXXVariantType(pub(crate) VariantType);

unsafe impl ExternType for CXXVariantType {
    type Id = type_id!("godot::ffi::VariantType");
    type Kind = cxx::kind::Trivial;
}

impl From<&ffi::Variant> for i64 {
    fn from(variant: &Variant) -> Self {
        ffi::variant_as_i64(variant)
    }
}

impl From<&ffi::Variant> for String {
    fn from(variant: &Variant) -> Self {
        ffi::variant_as_string(variant)
    }
}

impl From<&ffi::Variant> for bool {
    fn from(variant: &Variant) -> Self {
        ffi::variant_as_bool(variant)
    }
}

impl From<&ffi::Variant> for f64 {
    fn from(variant: &Variant) -> Self {
        ffi::variant_as_f64(variant)
    }
}
